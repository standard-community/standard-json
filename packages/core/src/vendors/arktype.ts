import type { Type } from "arktype";
import type { JSONSchema7 } from "json-schema";
import type { ToJsonSchemaFn } from "./index.js";

export const getToJsonSchemaFn =
  async (): Promise<ToJsonSchemaFn> => (schema) =>
    (schema as Type).toJsonSchema() as JSONSchema7;
