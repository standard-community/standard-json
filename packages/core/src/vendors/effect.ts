import type { Schema } from "effect";
import type { JSONSchema7 } from "json-schema";
import { tryImport, type ToJsonSchemaFn } from "./index.js";

export const getToJsonSchemaFn = async (): Promise<ToJsonSchemaFn> => {
  const { JSONSchema } = await tryImport(import("effect"), "effect");
  return (schema) =>
    JSONSchema.make(schema as Schema.Schema<unknown>) as JSONSchema7;
};
