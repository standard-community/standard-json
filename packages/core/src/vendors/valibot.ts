import type { JSONSchema7 } from "json-schema";
import { toJsonSchema as vToJsonSchema } from "@valibot/to-json-schema";
import type { BaseSchema } from "valibot";

export const toJsonSchema = (
  schema: BaseSchema<any, any, any>,
  options?: Record<string, unknown>,
) => vToJsonSchema(schema, options) as JSONSchema7;
