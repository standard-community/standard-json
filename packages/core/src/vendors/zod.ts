import type { JSONSchema7 } from "json-schema";
import { zodToJsonSchema } from "zod-to-json-schema";
import type * as z from "zod";

export const toJsonSchema = (
  schema: z.ZodType,
  options?: Record<string, unknown>,
) => zodToJsonSchema(schema, options) as JSONSchema7;
