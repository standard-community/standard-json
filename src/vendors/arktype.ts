import type { Type } from "arktype";
import type { JSONSchema7 } from "json-schema";
import type { ToJsonSchemaFn } from "./utils.js";

export default async function getToJsonSchemaFn(): Promise<ToJsonSchemaFn> {
  return (schema, options) =>
    (schema as Type).toJsonSchema(options) as JSONSchema7;
}
