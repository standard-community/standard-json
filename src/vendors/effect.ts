import type { Schema } from "effect";
import type { JSONSchema7 } from "json-schema";
import { errorMessageWrapper, type ToJsonSchemaFn } from "./utils.js";

export default async function getToJsonSchemaFn(): Promise<ToJsonSchemaFn> {
  const { JSONSchema } = await import("effect").catch(() => {
    throw new Error(errorMessageWrapper('Missing dependencies "effect".'));
  });
  return (schema) =>
    JSONSchema.make(schema as Schema.Schema<unknown>) as JSONSchema7;
}
