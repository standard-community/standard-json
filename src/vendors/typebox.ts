import type { JSONSchema7 } from "json-schema";
import type { Validator } from "typebox/compile";
import type { ToJsonSchemaFn } from "./utils.js";

export default function getToJsonSchemaFn(): ToJsonSchemaFn {
  return (schema) =>
    JSON.parse(JSON.stringify((schema as Validator).Type())) as JSONSchema7;
}
