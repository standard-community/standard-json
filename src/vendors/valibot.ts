import { errorMessageWrapper, type ToJsonSchemaFn } from "./utils.js";

export default async function getToJsonSchemaFn(): Promise<ToJsonSchemaFn> {
  try {
    const { toJsonSchema } = await import("@valibot/to-json-schema");
    return toJsonSchema as ToJsonSchemaFn;
  } catch {
    throw new Error(
      errorMessageWrapper('Missing dependencies "@valibot/to-json-schema".'),
    );
  }
}
