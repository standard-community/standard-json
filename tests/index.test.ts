import { type } from "arktype";
import { Schema } from "effect";
import Type from "typebox";
import { Compile } from "typebox/compile";
import * as v from "valibot";
import { describe, expect, it } from "vitest";
import { z } from "zod";
import { z as z4 } from "zod/v4";
import { z as zm } from "zod/v4-mini";

import { toJsonSchema } from "~/index.js";

describe("basic", () => {
  it("arktype", async () => {
    const schema = type({
      myString: "string",
      myUnion: "number | boolean",
    }).describe("My neat object schema");

    const jsonSchema = await toJsonSchema(schema);
    expect(jsonSchema).toMatchSnapshot();
  });

  it("effect", async () => {
    const schema = Schema.Struct({
      myString: Schema.String,
      myUnion: Schema.Union(Schema.Number, Schema.Boolean),
    });

    const jsonSchema = await toJsonSchema(Schema.standardSchemaV1(schema));
    expect(jsonSchema).toMatchSnapshot();
  });

  it("typebox", async () => {
    const schema = Type.Object(
      {
        myString: Type.String(),
        myUnion: Type.Union([Type.Number(), Type.Boolean()]),
      },
      { description: "My neat object schema" },
    );

    const jsonSchema = await toJsonSchema(Compile(schema));

    expect(jsonSchema).toMatchSnapshot();
  });

  it("valibot", async () => {
    const schema = v.pipe(
      v.object({
        myString: v.string(),
        myUnion: v.union([v.number(), v.boolean()]),
      }),
      v.description("My neat object schema"),
    );

    const jsonSchema = await toJsonSchema(schema);
    expect(jsonSchema).toMatchSnapshot();
  });

  it("zod", async () => {
    const zod4Schema = z4
      .object({
        myString: z4.string(),
        myUnion: z4.union([z4.number(), z4.boolean()]),
      })
      .describe("My neat object schema");

    const zod4JsonSchema = await toJsonSchema(zod4Schema);
    expect(zod4JsonSchema).toMatchSnapshot();

    const zod3Schema = z
      .object({
        myString: z.string(),
        myUnion: z.union([z.number(), z.boolean()]),
      })
      .describe("My neat object schema");

    const zod3JsonSchema = await toJsonSchema(zod3Schema);
    expect(zod3JsonSchema).toMatchSnapshot();

    const zodMiniSchema = zm.object({
      myString: zm.string(),
      myUnion: zm.union([zm.number(), zm.boolean()]),
    });

    const zodMiniJsonSchema = await toJsonSchema(zodMiniSchema);
    expect(zodMiniJsonSchema).toMatchSnapshot();
  });
});
