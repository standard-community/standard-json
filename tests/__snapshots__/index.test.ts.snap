// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > arktype 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`basic > effect 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`basic > valibot 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`basic > zod 1`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "additionalProperties": false,
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`basic > zod 2`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "type": [
        "number",
        "boolean",
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`basic > zod 3`] = `
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "additionalProperties": false,
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;
